Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMapping.cs

```diff
@@ -140,43 +140,79 @@ internal void SetDefaultInteractionMapping(bool overwrite = false)
+                    interactions[i] = new MixedRealityInteractionMapping(currentDefaultMapping)
+                    {
+                        MixedRealityInputAction = currentMapping.MixedRealityInputAction
+                    };
+
+                    updatedMappings = true;
                 }
+            }
 
-                for (int i = 0; i < newDefaultInteractions.Length; i++)
+            return updatedMappings;
+        }
+
+        private MixedRealityInteractionMapping[] CreateNewMatchedMapping(MixedRealityInteractionMapping[] interactions, MixedRealityInteractionMapping[] newDefaultInteractions)
+        {
+            MixedRealityInteractionMapping[] newDefaultMapping = new MixedRealityInteractionMapping[newDefaultInteractions.Length];
+
+            for (int i = 0; i < newDefaultInteractions.Length; i++)
+            {
+                for (int j = 0; j < interactions.Length; j++)
                 {
-                    MixedRealityInteractionMapping currentMapping = interactions[i];
-                    MixedRealityInteractionMapping currentDefaultMapping = newDefaultInteractions[i];
-
-                    if (currentMapping.Id != currentDefaultMapping.Id ||
-                        currentMapping.Description != currentDefaultMapping.Description ||
-                        currentMapping.AxisType != currentDefaultMapping.AxisType ||
-                        currentMapping.InputType != currentDefaultMapping.InputType ||
-                        currentMapping.KeyCode != currentDefaultMapping.KeyCode ||
-                        currentMapping.AxisCodeX != currentDefaultMapping.AxisCodeX ||
-                        currentMapping.AxisCodeY != currentDefaultMapping.AxisCodeY ||
-                        currentMapping.InvertXAxis != currentDefaultMapping.InvertXAxis ||
-                        currentMapping.InvertYAxis != currentDefaultMapping.InvertYAxis)
+                    if (Equals(interactions[j], newDefaultInteractions[i]))
                     {
-                        interactions[i] = new MixedRealityInteractionMapping(currentDefaultMapping)
+                        newDefaultMapping[i] = new MixedRealityInteractionMapping(newDefaultInteractions[i])
                         {
-                            MixedRealityInputAction = currentMapping.MixedRealityInputAction
+                            MixedRealityInputAction = interactions[j].MixedRealityInputAction
                         };
-
-                        updatedMappings = true;
+                        break;
                     }
                 }
+
+                if (newDefaultMapping[i] == null)
+                {
+                    newDefaultMapping[i] = new MixedRealityInteractionMapping(newDefaultInteractions[i]);
+                }
             }
 
-            return updatedMappings;
+            return newDefaultMapping;
+        }
+
+        private bool Equals(MixedRealityInteractionMapping a, MixedRealityInteractionMapping b)
```

Comment:
Do we want to override `MixedRealityInteractionMapping`'s `.Equals()` method instead?

Reply:
If so, we can add that later and in a compatible way. When we do, we should consider overriding == and != as well

Reply:
> If so, we can add that later and in a compatible way

What do you mean by that?